go/teams/audit.go

```diff
@@ -1,15 +1,409 @@
+	defer m.CTrace("Auditor#doPostProbes", func() error { return err })()
+
+	var low keybase1.Seqno
+	last := lastAudit(history)
+	var prev *probeTuple
+	if last == nil {
+		low = headMerkle.Seqno
+	} else {
+		low = last.MaxMerkleSeqno
+		probe, ok := history.PostProbes[last.MaxMerkleProbe]
+		if !ok {
+			return maxMerkleProbe, NewAuditError("previous audit pointed to a bogus probe (seqno=%d)", last.MaxMerkleProbe)
+		}
+		prev = &probeTuple{
+			merkle: last.MaxMerkleProbe,
+			team:   probe.TeamSeqno,
+			// leave linkID nil, it's not needed...
+		}
+	}
+
+	probeTuples, err := a.computeProbes(m, history.ID, history.PostProbes, probeID, low, latestMerkleSeqno, 0, params.NumPostProbes)
+	if err != nil {
+		return maxMerkleProbe, err
+	}
+	if len(probeTuples) == 0 {
+		m.CDebugf("No probe tuples, so bailing")
+		return maxMerkleProbe, nil
+	}
+
+	for _, tuple := range probeTuples {
+		m.CDebugf("postProbe: checking probe at %+v", tuple)
+		if tuple.team > keybase1.Seqno(0) {
+			expectedLinkID, ok := chain[tuple.team]
+			if !ok {
+				return maxMerkleProbe, NewAuditError("team chain doesn't have a link for seqno %d, but expected one", tuple.team)
+			}
+			if !expectedLinkID.Eq(tuple.linkID) {
+				return maxMerkleProbe, NewAuditError("team chain linkID mismatch at %d: wanted %s but got %s via merkle seqno %d", tuple.team, expectedLinkID, tuple.linkID, tuple.merkle)
+			}
+		}
+		history.PostProbes[tuple.merkle] = keybase1.Probe{Index: probeID, TeamSeqno: tuple.team}
+		if prev != nil && prev.team > tuple.team {
+			return maxMerkleProbe, NewAuditError("team chain unexpected jump: %d > %d via merkle seqno %d", prev.team, tuple.team, tuple.merkle)
+		}
+		if tuple.merkle > maxMerkleProbe {
+			maxMerkleProbe = tuple.merkle
+		}
+		prev = &tuple
+	}
+	return maxMerkleProbe, nil
+}
+
+func (a *Auditor) doPreProbes(m libkb.MetaContext, history *keybase1.AuditHistory, probeID int, headMerkle keybase1.MerkleRootV2) (err error) {
+	defer m.CTrace("Auditor#doPreProbes", func() error { return err })()
+
+	first := m.G().MerkleClient.FirstSeqnoWithSkips(m)
+	if first == nil {
+		return NewAuditError("cannot find a first modern merkle sequence")
+	}
+
+	probeTuples, err := a.computeProbes(m, history.ID, history.PreProbes, probeID, *first, headMerkle.Seqno, len(history.PreProbes), params.NumPreProbes)
+	if err != nil {
+		return err
+	}
+	if len(probeTuples) == 0 {
+		m.CDebugf("No probe pairs, so bailing")
+		return nil
+	}
+	for _, tuple := range probeTuples {
+		m.CDebugf("preProbe: checking probe at merkle %d", tuple.merkle)
+		if tuple.team != keybase1.Seqno(0) || !tuple.linkID.IsNil() {
+			return NewAuditError("merkle root at %v should have been nil for %v; got %s/%d",
+				tuple.merkle, history.ID, tuple.linkID, tuple.team)
+		}
+	}
+	return nil
+}
+
+func randSeqno(lo keybase1.Seqno, hi keybase1.Seqno) (keybase1.Seqno, error) {
+	rng := hi - lo + 1
+	rngBig := big.NewInt(int64(rng))
+	n, err := rand.Int(rand.Reader, rngBig)
```

Comment:
If we're using for security this should be a CSPRNG

Reply:
thanks for checking this, though I believe it is!